pubsub/awssnssqs/awssnssqs.go

```diff
@@ -411,21 +419,55 @@ func OpenSubscription(ctx context.Context, client *sqs.SQS, qURL string, opts *S
 
 // openSubscription returns a driver.Subscription.
 func openSubscription(ctx context.Context, client *sqs.SQS, qURL string) driver.Subscription {
-	return &subscription{client: client, qURL: qURL}
+	return &subscription{
+		client: client,
+		qURL:   qURL,
+		acksem: make(chan struct{}, maxAckConcurrency),
+	}
 }
 
-// How long ReceiveBatch should wait if no messages are available; controls
-// the poll interval of requests to SQS.
-const noMessagesPollDuration = 250 * time.Millisecond
-
 // ReceiveBatch implements driver.Subscription.ReceiveBatch.
-func (s *subscription) ReceiveBatch(ctx context.Context, maxMessages int) (msgs []*driver.Message, er error) {
+func (s *subscription) ReceiveBatch(ctx context.Context, maxMessages int) ([]*driver.Message, error) {
 	// SQS supports receiving at most 10 messages at a time:
 	// https://docs.aws.amazon.com/AWSSimpleQueueService/latest/APIReference/API_ReceiveMessage.html
-	max := int64(maxMessages)
-	if max > 10 {
-		max = 10
+	// If maxMessages is larger than 10, we make up to maxReceiveConcurrency RPCs
+	// in parallel to try to get more.
+	const maxMessagesPerRPC = 10
+
+	var mu sync.Mutex
+	var ms []*driver.Message
+
+	g, ctx := errgroup.WithContext(ctx)
+	for n := 0; n < maxReceiveConcurrency && maxMessages > 0; n++ {
```

Comment:
This will behave well when lots of messages are available. But if there's only a few, you'll be making 100 RPCs, most of which will return nothing. Does Amazon charge per RPC, or only per message? I'm also worried about quota.

Could you add a TODO? One alternative is to remember how many messages you got the last time, and tailor the number of calls to that. We could even use a weighted moving average.

Reply:
Hmm, not sure I understand. This should only make `maxMessages / 10` RPCs at most, with a max of 100. So, e.g., at startup when `maxMessages = 1`, it will only make 1 RPC.

And after that `maxMessages` will only increase if the throughput of messages is high. While it's possible that 
-- we're going through a huge number of messages/sec
-- `maxMessages` ramps up to >= 1000
-- suddenly there aren't any more messages on the server (or many fewer)
and we'll make a bunch of unnecessary RPCs, that should be limited in time because `maxMessages` will immediately start to ramp down again.

In other words, I think that the `maxMessages` argument does what you suggested in the TODO.

Reply:
Yes, I see.